### PR TITLE
Add support of non exists commits

### LINKED ISSUE #20772: Running a code insight over a timeframe when the repo didn't yet exist in Sourcegraph breaks the entire insight, but it should just not display data for that date(s)
- **Sourcegraph version:**  3.27 and earlier

#### Steps to reproduce:

1. Go to any code insight in dotcom settings
2. Change the step to be something large, like `"years": 2` or similar (so shows 14 years on the chart) 
3. View the insight and you'll see: 
image

The cause of this bug is that `bulkSearchCommits` doesn't find a commit, and then it breaks the insight when it runs the query over the commits returned from bulkSearchCommits (because there's no commit that existed). 

#### Expected behavior:

**We should** modify so that instead of breaking the insight, the graph should just show no points if we don't find a commit. This means it may only show 3-4 points; that's fine. We should keep the timeframe x-axis the same length to illustrate this. (Strong hypothesis: most folks, if they see a graph that starts midway along the x-axis, will understand it's due to lack of data before that). 

**Additional functionality if time or for later iterations:** for the earliest-in-time commit that we do have, we should include that to demonstrate that's the starting point of the graph. 

> For example, let's say I have a repo that only has commits synced to Sourcegraph dating back to August 2017, and I want a code insight that runs over the last 7 years starting now (May). I would see a datapoint for May '21, May '20, May '19, May '18, **Aug '17**, and then nothing else. 

**Caveat to the above: If there are multiple repos for the data series**, we should simply count "pre-repo-existed" datapoints as 0. If the time-axis pre-dates all repos in the insight, then we should do the "first datapoint" on the graph for the earliest time at least 1 repo existed, and otherwise continue to sum the results count from the other repos as 0 until data exists. 

> This means in the earlier example if 1 repo dates back to August 2017 and another dates back to November 2017 – and my graph marks points 1 per year in May (now) – then I'll see an august '17 data point from repo 1 (and no Nov 2017 datapoint), and then the May 2018 datapoint will include total result count from both repos. 

**Stretch goal (later iteration):** we should non-blocking warn you on insight creation if your timeframe extends past the date of your insight, just so you are aware, and suggest the earliest available timeframe that has data. 

#### Actual behavior:

It breaks the insight and shows nothing at all but the error message.

### PR DESCRIPTION
**Pull request by vovakulikov**
Close https://github.com/sourcegraph/sourcegraph/issues/20772

**Important** This PR should be merged together with https://github.com/sourcegraph/sourcegraph/pull/21288

This PR changes fetching logic in a way to support partial commits loading in case if a user sets a big-time step (years) for their insight and got the moment on the chart when a particular repository hasn't been created yet. 

Now we just skip these data points and show data that matter from the repo lifetime.

image

### CODE REVIEW

File: package.json
Comment (vovakulikov) [inline package.json:158]: I also added prettier and prettier script to align code style in this repo

**Comment by felixfbecker** [comment]
Two questions I have:
1. Does this change the return value/type of the extension in any way?
2. Is this change backwards-compatible with older Sourcegraph versions, i.e. this new extension version will continue to work on old instances?

**Comment by vovakulikov** [comment]
@felixfbecker answered these questions here
https://github.com/sourcegraph/sourcegraph/pull/21288#issuecomment-849682928
But this thread seems more acceptable for this discussion about extension API so

>Does this change the return value/type of the extension in any way?

Yes and no, we've changed datum value but from types perspective, nothing has been changed since we have datum as a generic in our types

```ts
 export interface LineChartContent<D extends object, XK extends keyof D> {
        chart: 'line'

        /** An array of data objects, with one element for each step on the X-axis. */
        data: D[]

        /** The series (lines) of the chart. */
        series: {}[]

        xAxis: ChartAxis<XK, D>
    }
```

>Is this change backward-compatible with older Sourcegraph versions, i.e. this new extension version will continue to work on old instances?

The null value is a valid data point for our current `visx` charts but I'm not sure about old charts (re-chart library) I believe we still have these charts in sourcegraph 3.26. Do you have any suggestions on how can  we publish a major update of extension for sourcegraph 3.27 and higher?

**Comment by felixfbecker** [comment]
Major updates don't exist for extensions, it's always just one latest version that needs to be backwards-compatible to roughly the last 2 versions.

If this turns out to cause errors, what if we just returned `0` here (which feels semantically correct) and change it to `null` in a future release?

**Comment by vovakulikov** [comment]
@felixfbecker I've thought about it, but using zero doesn't seem to be enough. The problem here is how can we separate situations when series got `bad commits` and then we should render non-active background and when series got valid commits but with zero values and we should not render non-active background but actually SVG line with zero values.

Also using zero values for bad commits is bad because we can break the visual representation of the chart in case if other series have a big Y-value. All series would be collapsed if we add zero value to one of the other series.

**Comment by vovakulikov** [comment]
>needs to be backward-compatible to roughly the last 2 versions.

it seems like these changes are suitable for this policy. Our re-charts insights were available in 3.26. We released new charts in 3.27 so 3.27 and 3.28 work with these major null-changes here.

But I understand the possible problem here. @felixfbecker wdyt perhaps we can add an experimental flag for this new behavior with null and enable it by default. If our customers have any problems with the backward compatibility of this extension, they can disable this logic.

**Comment by vovakulikov** [comment]
@felixfbecker I checked re-charts with null values and  zero values do not break the line charts of re-chart so it seems we are completely backward compatible with 3.26 

https://codesandbox.io/s/simple-line-chart-forked-8ze5k

**Comment by felixfbecker** [comment]
Awesome! I have no issues then and I don't think we need a flag. But we should wait for a code review from @valerybugakov (or someone else, if he can't get to it right now)

**Review by eseliger — APPROVED** [review]

**Review by valerybugakov — APPROVED** [review]
LGTM! Thanks for improving the infrastructure of this repo.
One note: it would easier to review if this PR has two separate commits:
1. For the infra upgrade.
2. For the actual logic change.

We're guilty of the same thing in the frontend platform team :)
Please, ping us if you see any of our PRs/commits could be split further to ease the review process.

**Pull request merged**
